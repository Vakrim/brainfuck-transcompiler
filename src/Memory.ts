import { Address } from "./Address";
import { Scope } from "./Scope";

export class Memory {
  #tape: Map<Address, Scope>;
  #dirties: Set<Address>;

  constructor() {
    this.#tape = new Map();
    this.#dirties = new Set();
  }

  allocate(scope: Scope, nextTo: Address): Allocation {
    const address = this.#findFreeAdrress(nextTo);
    const isDirty = this.#dirties.has(address);

    this.#tape.set(address, scope);
    this.#dirties.delete(address);

    return { address, isDirty };
  }

  free(address: Address, isDirty: boolean, scope: Scope) {
    const cell = this.#tape.get(address);
    if (!cell) {
      throw new Error(`Cell with address = ${address} is not allocated`);
    }
    if (cell !== scope) {
      throw new Error(
        `Cell with address = ${address} is not allocated by the scope that is requesting to free it`
      );
    }

    if (isDirty) {
      this.#dirties.add(address);
    }
    this.#tape.delete(address);
  }

  #findFreeAdrress(nextTo: Address) {
    for (let searchRange = 0; ; searchRange++) {
      if (
        nextTo - searchRange >= 0 &&
        !this.#tape.has((nextTo - searchRange) as Address)
      ) {
        return (nextTo - searchRange) as Address;
      }

      if (!this.#tape.has((nextTo + searchRange) as Address)) {
        return (nextTo + searchRange) as Address;
      }
    }
  }

  get dirties() {
    return [...this.#dirties];
  }
}

interface Allocation {
  address: Address;
  isDirty: boolean;
}
