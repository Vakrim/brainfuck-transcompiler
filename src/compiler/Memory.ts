import { Address } from './Address';

export class Memory {
  #tape: Map<Address, ScopeLike>;

  constructor() {
    this.#tape = new Map();
  }

  allocate(scope: ScopeLike, nextTo: Address, size: number): Address {
    const address = this.#findFreeAdrress(nextTo, size);

    for (let i = 0; i < size; i++) {
      this.#tape.set((address + i) as Address, scope);
    }

    return address;
  }

  free(scope: ScopeLike, address: Address) {
    const cell = this.#tape.get(address);
    if (!cell) {
      throw new Error(`Cell with address = ${address} is not allocated`);
    }
    if (cell !== scope) {
      throw new Error(
        `Cell with address = ${address} is not allocated by the scope that is requesting to free it`
      );
    }

    this.#tape.delete(address);
  }

  #findFreeAdrress(nextTo: Address, size: number) {
    for (let searchRange = 0; ; searchRange++) {
      for (let i = 0; i < size; i++) {
        if (
          nextTo - searchRange - i < 0 ||
          this.#tape.has((nextTo - searchRange - i) as Address)
        ) {
          break;
        }
        if (i >= size - 1) {
          return (nextTo - searchRange - size + 1) as Address;
        }
      }

      for (let i = 0; i < size; i++) {
        if (this.#tape.has((nextTo + searchRange + i) as Address)) {
          break;
        }
        if (i >= size - 1) {
          return (nextTo + searchRange) as Address;
        }
      }
    }
  }
}

type ScopeLike = object;
