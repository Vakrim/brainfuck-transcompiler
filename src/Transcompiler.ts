import { Address } from "./Address";
import { Memory } from "./Memory";
import { Scope } from "./Scope";
import { TemporaryVariable } from "./TemporaryVariable";
import { Variable } from "./Variable";

export class Transcompiler {
  #cursorPosition: Address = 0 as Address;
  #memory: Memory;
  #scope: Scope;
  #code: string;
  #commentBuffer: string;

  constructor(memory = new Memory()) {
    this.#cursorPosition = 0 as Address;
    this.#memory = memory;
    this.#scope = new Scope(null, this.#memory);
    this.#code = "";
    this.#commentBuffer = "";
  }

  declareVariable(name: string) {
    this.comment(`declare ${name}`);

    const isDirty = this.#scope.declareVariable(name, this.#cursorPosition);
    if (isDirty) {
      this.#reset(name);
    }
  }

  assignValue(name: string, value: number) {
    this.comment(`assign ${value} to ${name}`);

    this.#moveToVariable(name);
    this.#setValue(value);
  }

  writeInput(name: string) {
    this.comment(`write input ${name}`);

    this.#moveToVariable(name);
    this.#outputBrainfuck(",");
  }

  readVariable(name: string) {
    this.comment(`read ${name}`);

    this.#moveToVariable(name);
    this.#readValue();
  }

  add(to: string, from: string) {
    this.comment(`add ${from} to ${to}`);

    this.#moveToVariable(from);
    const newFrom = this.#declareTemporaryVariable(this.#cursorPosition);

    this.#loopOf(from, () => {
      this.#moveToVariable(newFrom);
      this.#inc();

      this.#moveToVariable(to);
      this.#inc();
    });

    this.#scope.promoteVariable(from, newFrom);
  }

  multiply(to: string, from: string) {
    this.comment(`multiply ${from} to ${to}`);

    const interator = this.#copy(from);
    const interator2 = this.#copy(to);
    const interator2swap = this.#declareTemporaryVariable(interator.address);
    const result = this.#declareTemporaryVariable(interator.address);

    this.#loopOf(interator, () => {
      this.#loopOf(interator2, () => {
        this.#moveToVariable(interator2swap);
        this.#inc();
        this.#moveToVariable(result);
        this.#inc();
      });
      this.#loopOf(interator2swap, () => {
        this.#moveToVariable(interator2);
        this.#inc();
      });
    });

    this.#scope.unsetTemporaryVariable(interator);
    this.#scope.unsetTemporaryVariable(interator2);
    this.#scope.unsetTemporaryVariable(interator2swap);
    this.#scope.promoteVariable(to, result);
  }

  pushScope() {
    this.#scope = new Scope(this.#scope, this.#memory);
  }

  popScope() {
    const variables = this.#scope.getVariablesOfThisScope();
    for (const variable of variables) {
      this.#scope.unsetVariable(variable);
    }
    this.#scope = this.#scope.getParentScope();
  }

  scope(fn: () => void) {
    this.pushScope();
    fn();
    this.popScope();
  }

  comment(comment: string) {
    const safeComment = comment.replace(/[\,\.\[\]\<\>\+\-]/, " ");

    this.#commentBuffer = this.#commentBuffer
      ? [this.#commentBuffer, safeComment].join("; ")
      : safeComment;
  }

  get code() {
    return `${this.#code}\n${
      this.#commentBuffer ? `${this.#commentBuffer}\n` : ""
    }`;
  }

  #declareTemporaryVariable(nextTo: Address = this.#cursorPosition) {
    const { variable, isDirty } = this.#scope.declareTemporaryVariable(nextTo);
    if (isDirty) {
      this.#reset(variable);
    }
    return variable;
  }

  #copy(from: string) {
    this.comment(`copy ${from} to temp`);

    const newFrom = this.#declareTemporaryVariable();
    const newTo = this.#declareTemporaryVariable();

    this.#loopOf(from, () => {
      this.#moveToVariable(newFrom);
      this.#inc();
      this.#moveToVariable(newTo);
      this.#inc();
    });

    this.#scope.promoteVariable(from, newFrom);

    return newTo;
  }

  #moveToVariable(nameOrVariable: VariableLike) {
    this.#moveTo(this.#normalizeVariable(nameOrVariable).address);
  }

  #reset(name: VariableLike) {
    this.#moveToVariable(name);
    this.#outputBrainfuck("[-]");
  }

  #normalizeVariable(variableLike: VariableLike) {
    if (typeof variableLike === "string") {
      return this.#scope.getVariable(variableLike);
    }
    return variableLike;
  }

  #moveTo(address: Address) {
    const difference = address - this.#cursorPosition;
    if (difference < 0) {
      this.#outputBrainfuck("<".repeat(-difference));
    } else if (difference > 0) {
      this.#outputBrainfuck(">".repeat(difference));
    }
    this.#cursorPosition = address;
  }

  #loopOf(variableLike: VariableLike, fn: () => void) {
    const address = this.#normalizeVariable(variableLike).address;

    this.#moveTo(address);
    this.#outputBrainfuck(`[`);
    this.#dec();
    fn();
    this.#moveTo(address);
    this.#outputBrainfuck(`]`);
  }

  #outputBrainfuck(bf: string) {
    if (this.#commentBuffer) {
      this.#code += `\n${this.#commentBuffer}\n`;
      this.#commentBuffer = "";
    }
    this.#code += bf;
  }

  #dec() {
    this.#outputBrainfuck(`-`);
  }

  #inc() {
    this.#outputBrainfuck(`+`);
  }

  #setValue(value: number) {
    this.#outputBrainfuck("+".repeat(value));
  }

  #readValue() {
    this.#outputBrainfuck(".");
  }
}

type VariableLike = string | Variable | TemporaryVariable;
