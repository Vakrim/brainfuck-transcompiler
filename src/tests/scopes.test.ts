import { Memory } from '../Memory';
import { Transcompiler } from '../Transcompiler';
import { executeCode } from './execute-code';

describe('scopes', () => {
  it('creates scope assign variable', () => {
    const memory = new Memory();
    const allocateSpy = jest.spyOn(memory, 'allocate');
    const freeSpy = jest.spyOn(memory, 'free');

    const compiler = new Transcompiler(memory);

    compiler.declareVariable('outerScope');

    expect(allocateSpy).toHaveBeenCalledTimes(1);
    expect(freeSpy).toHaveBeenCalledTimes(0);

    compiler.pushScope();

    compiler.declareVariable('innerScope');
    compiler.assignValue('innerScope', 3);
    compiler.readVariable('innerScope');

    expect(allocateSpy).toHaveBeenCalledTimes(2);
    expect(freeSpy).toHaveBeenCalledTimes(0);

    compiler.popScope();

    expect(allocateSpy).toHaveBeenCalledTimes(2);
    expect(freeSpy).toHaveBeenCalledTimes(1);

    compiler.declareVariable('outerScopeInPlace');
    compiler.assignValue('outerScopeInPlace', 5);
    compiler.readVariable('outerScopeInPlace');

    expect(allocateSpy).toHaveBeenCalledTimes(3);
    expect(freeSpy).toHaveBeenCalledTimes(1);

    expect(compiler.code).toMatchSnapshot();

    expect(executeCode(compiler).codes).toEqual([3, 5]);
  });

  it('assign memory after inner scope when usign temporary variables in outer scope', () => {
    const compiler = new Transcompiler();

    compiler.declareVariable('a');
    compiler.assignValue('a', 3);

    compiler.pushScope();
    compiler.declareVariable('b');
    compiler.assignValue('b', 5);

    compiler.add('b', 'a');
    compiler.readVariable('b');
    compiler.readVariable('a');

    compiler.popScope();

    expect(compiler.code).toMatchSnapshot();

    expect(executeCode(compiler).codes).toEqual([8, 3]);
  });
});
