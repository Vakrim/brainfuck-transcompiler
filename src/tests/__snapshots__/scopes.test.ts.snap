// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scopes assign memory after inner scope when usign temporary variables in outer scope 1`] = `
"
declare a; assign 3 to a
+++
declare b; assign 5 to b
>+++++
add a to b
<[->>+<+<]
read b
>.
read a
>.
"
`;

exports[`scopes creates scope assign variable 1`] = `
"
declare outerScope; declare innerScope; assign 3 to innerScope
>+++
read innerScope
.
declare outerScopeInPlace
[-]
assign 5 to outerScopeInPlace
+++++
read outerScopeInPlace
.
"
`;
